homeassistant/components/mqtt/humidifier.py

```diff
@@ -0,0 +1,458 @@
+    CONF_RETAIN,
+    CONF_STATE_TOPIC,
+    DOMAIN,
+    PLATFORMS,
+    subscription,
+)
+from .. import mqtt
+from .debug_info import log_messages
+from .mixins import MQTT_ENTITY_COMMON_SCHEMA, MqttEntity, async_setup_entry_helper
+
+CONF_AVAILABLE_MODES_LIST = "modes"
+CONF_COMMAND_TEMPLATE = "command_template"
+CONF_DEVICE_CLASS = "device_class"
+CONF_MODE_COMMAND_TEMPLATE = "mode_command_template"
+CONF_MODE_COMMAND_TOPIC = "mode_command_topic"
+CONF_MODE_STATE_TOPIC = "mode_state_topic"
+CONF_MODE_STATE_TEMPLATE = "mode_state_template"
+CONF_PAYLOAD_RESET_MODE = "payload_reset_mode"
+CONF_PAYLOAD_RESET_HUMIDITY = "payload_reset_humidity"
+CONF_STATE_VALUE_TEMPLATE = "state_value_template"
+CONF_TARGET_HUMIDITY_COMMAND_TEMPLATE = "target_humidity_command_template"
+CONF_TARGET_HUMIDITY_COMMAND_TOPIC = "target_humidity_command_topic"
+CONF_TARGET_HUMIDITY_MIN = "min_humidity"
+CONF_TARGET_HUMIDITY_MAX = "max_humidity"
+CONF_TARGET_HUMIDITY_STATE_TEMPLATE = "target_humidity_state_template"
+CONF_TARGET_HUMIDITY_STATE_TOPIC = "target_humidity_state_topic"
+
+DEFAULT_NAME = "MQTT Humidifier"
+DEFAULT_OPTIMISTIC = False
+DEFAULT_PAYLOAD_ON = "ON"
+DEFAULT_PAYLOAD_OFF = "OFF"
+DEFAULT_PAYLOAD_RESET = "None"
+
+MQTT_HUMIDIFIER_ATTRIBUTES_BLOCKED = frozenset(
+    {
+        humidifier.ATTR_HUMIDITY,
+        humidifier.ATTR_MAX_HUMIDITY,
+        humidifier.ATTR_MIN_HUMIDITY,
+        humidifier.ATTR_MODE,
+        humidifier.ATTR_AVAILABLE_MODES,
+    }
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def valid_mode_configuration(config):
+    """Validate that the mode reset payload is not one of the available modes."""
+    if config.get(CONF_PAYLOAD_RESET_MODE) in config.get(CONF_AVAILABLE_MODES_LIST):
+        raise ValueError("modes must not contain payload_reset_mode")
+    return config
+
+
+def valid_humidity_range_configuration(config):
+    """Validate that the target_humidity range configuration is valid, throws if it isn't."""
+    if config.get(CONF_TARGET_HUMIDITY_MIN) >= config.get(CONF_TARGET_HUMIDITY_MAX):
+        raise ValueError("target_humidity_max must be > target_humidity_min")
+    if config.get(CONF_TARGET_HUMIDITY_MAX) > 100:
+        raise ValueError("max_humidity must be <= 100")
+
+    return config
+
+
+PLATFORM_SCHEMA = vol.All(
+    mqtt.MQTT_RW_PLATFORM_SCHEMA.extend(
+        {
+            # CONF_AVAIALABLE_MODES_LIST and CONF_MODE_COMMAND_TOPIC must be used together
+            vol.Inclusive(
+                CONF_AVAILABLE_MODES_LIST, "available_modes", default=[]
+            ): cv.ensure_list,
+            vol.Inclusive(
+                CONF_MODE_COMMAND_TOPIC, "available_modes"
+            ): mqtt.valid_publish_topic,
+            vol.Optional(CONF_COMMAND_TEMPLATE): cv.template,
+            vol.Optional(
+                CONF_DEVICE_CLASS, default=DEVICE_CLASS_HUMIDIFIER
+            ): cv.matches_regex(
+                f"^({DEVICE_CLASS_HUMIDIFIER}|{DEVICE_CLASS_DEHUMIDIFIER})$"
+            ),
```

Comment:
Use voluptuous.In instead:

```suggestion
            ): vol.In([DEVICE_CLASS_HUMIDIFIER, DEVICE_CLASS_DEHUMIDIFIER]),
```